### PR TITLE
Algorithm #39 UnionFind WIP

### PR DESCRIPTION
**Pull request by mknblch**
This one is still work in progress please dont merge yet. I implemented a disjoint-set-struct which will be needed for some algorithms.

### CODE REVIEW

File: core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+
+    /**
+     * find setId of element p.
+     * @param p the element in the set we are looking for
+     * @return an id of the set it belongs to
+     */
+    public int find(int p) {
+        if (parent[p] == p) return p;
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java]: should we save repeated array access with local variables?

File: core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    private final int[] parent;
+    private final int[] depth;
+
+    /**
+     *
+     * @param capacity the capacity (maximum node id)
+     */
+    public DisjointSetStruct(int capacity) {
+        parent = new int[capacity];
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java]: can we handle this with less memory usage? e.g. with a bitset?

File: algo/src/main/java/org/neo4j/graphalgo/impl/UnionFind.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        return struct;
+    }
+
+    // TODO visibility
+    @Override
+    public void accept(int sourceNodeId, int targetNodeId, long relationId) {
+        if (!struct.connected(sourceNodeId, targetNodeId)) {
+            struct.union(sourceNodeId, targetNodeId);
+        } else {
+            System.out.println("cycle");
```
Comment (jexp) [inline algo/src/main/java/org/neo4j/graphalgo/impl/UnionFind.java]: probably record it somewhere

File: core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * find setId of element p.
+     * @param p the element in the set we are looking for
+     * @return an id of the set it belongs to
+     */
+    public int find(int p) {
+        if (parent[p] == p) return p;
+        // path compression optimization
+        parent[p] = find(parent[p]); // balance tree while traversing
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java]: recursion, can we have huge depths in larger graphs? perhaps replace with iteration? (i.e. a loop)

File: core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * join set of p (Sp) with set of q (Sq)
+     *
+     * @param p an item of Sp
+     * @param q an item of Sq
+     */
+    public void union(int p, int q) {
+        final int pSet = find(p);
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java]: potential check if p = q?

File: core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java (hunk 4 above)
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java]: You are true. I've seen 2 possible optimizations in the wild. One is path-halving which works nonrecursive but might at the end not be as efficient as this one. The other one is path compression by recursivly shortening the path during traversion. According to wiki we should get an overall complexity of O(a(n)) ( a(..) is Inverse-Ackermann function). Anyways.. I'd like to do some real-world tests and benchmars on both implementations.

File: algo/src/main/java/org/neo4j/graphalgo/impl/UnionFind.java (hunk 3 above)
Comment (mknblch) [inline algo/src/main/java/org/neo4j/graphalgo/impl/UnionFind.java]: Sure. This class is still 99% WIP.

File: core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java (hunk 2 above)
Comment (mknblch) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/DisjointSetStruct.java]: The parent array is more like an index-to-parentIndex map. We can switch to an implementation based on linked lists which could grow on-demand but as usual decrease performance. We might consider having both implementations and some kind of configuration to choose one.

**Comment by jexp** [comment]
Lemme know how far you got.

**Comment by mknblch** [comment]
delayed until api2 refactoring is done

**Comment by mknblch** [comment]
still WIP, please do not merge yet.

File: algo/src/main/java/org/neo4j/graphalgo/impl/MSTPrim.java [hunk 6]
```diff
@@ -28,63 +33,95 @@ public MSTPrim(Graph graph) {
...
+        public MinimumSpanningTree(int expectedElements) {
+            costMap = new LongDoubleScatterMap(expectedElements);
         }
 
-        /**
-         * get the element count in this mst
-         */
-        public int size() {
-            return offset;
+        public void add(int sourceNodeId, int targetNodeId, double cost) {
```
Comment (mknblch) [inline algo/src/main/java/org/neo4j/graphalgo/impl/MSTPrim.java]: We split the value into ints in line 57 and combine them again in 80. Change method signature (combinedSourceTarget:long, cost:double):void

File: tests/src/test/java/org/neo4j/graphalgo/impl/ThresholdUnionFindTest.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        a = idMapper.toMappedNodeId(n1);
+        b = idMapper.toMappedNodeId(n2);
+        c = idMapper.toMappedNodeId(n3);
+        d = idMapper.toMappedNodeId(n4);
+        e = idMapper.toMappedNodeId(n5);
+        f = idMapper.toMappedNodeId(n6);
+    }
+
+    /**
+     * should split the graph into 2 distinct sets
```
Comment (mknblch) [inline tests/src/test/java/org/neo4j/graphalgo/impl/ThresholdUnionFindTest.java]: 3

**Comment by mknblch** [comment]
can be merged. 

- unionFind clustering algorithm (#39) + several data structures + test + benchmark
- changes the ConsumerIFaces according to #68
- implements Weights as datasource #73

File: core/src/main/java/org/neo4j/graphalgo/core/utils/RawValues.java [hunk 8]
```diff
@@ -35,4 +35,32 @@ public static int getHead(long combinedValue) {
...
+
+    /**
+     * convert property value to double
+     *
+     * @param value the value object
+     * @param propertyDefaultValue default value if property cant be converted
+     * @return double representation of value
+     */
+    public static double extractValue(Object value, double propertyDefaultValue) {
+        if (value instanceof Number) {
```
Comment (jexp) [inline core/src/main/java/org/neo4j/graphalgo/core/utils/RawValues.java]: probably add an early null check for returning default values.

**Pull request merged**